File: snippets/TreeLongestPath.java
public class TreeLongestPath {
	
	private static Pair<Integer, Integer> longestPath(Tree tree) {
		if (tree.children == null) {
			return new Pair<Integer, Integer>(0, 0);
		}
		int childMax = 0, mh1 = -1, mh2 = -1;
		for (Tree child : tree.children) {
			Pair<Integer, Integer> m = longestPath(child);
			if (m.second > childMax) {
				childMax = m.second;
			}
			if (m.first > mh1) {
				mh2 = mh1;
				mh1 = m.first;
			}
			else if (m.first > mh2) {
				mh2 = m.first;
			}
		}
		return new Pair<Integer, Integer>(Math.max(mh1, mh2) + 1, Math.max(childMax, mh1 + mh2 + 2));
	}

	private static int height(Tree tree) {
		if (tree.children == null) {
			return 1;
		}
		int cmh = 0;
		for (Tree child : tree.children) {
			int ch = height(child); 
			if (ch > cmh) {
				cmh = ch;
			}
		}
		return cmh + 1;
	}

	public static void main(String[] args) {
		Tree tree = TreeBuilder.getSampleTree();
		Pair<Integer, Integer> max = longestPath(tree);
		System.out.println(String.format("Heigth: %d. Longest Path: %d", height(tree), Math.max(max.first, max.second)));
	}

}

File: snippets/TreeLongestPath.cs
using System;
using System.Collections.Generic;

namespace CRClient
{
	class TreeLongestPath
	{
		public static void Main(string[] args)
		{
			var tree = BuildSampleTree();
			var lp = LongestPath(tree);
			Console.WriteLine(lp.Item2);
			Console.ReadLine();
		}

		private static Tree BuildSampleTree()
		{
			return new Tree(1, new[] { new Tree(2), new Tree(3, new[] { new Tree(5) }), new Tree(4) });
		}

		/// Returns a tuple with values, first: Height of the tree rooted at tree and second: longest path of // any tree rooted at tree
		private static Tuple<int, int> LongestPath(Tree tree)
		{
			if (tree.Children == null)
			{
				return new Tuple<int, int>(0, 0);
			}
			int mh1 = -1, mh2 = -1, childMax = 0;
			foreach (var child in tree.Children)
			{
				var clp = LongestPath(child);
				if (clp.Item2 > childMax)
				{
					childMax = clp.Item2;
				}
				if (clp.Item1 > mh1)
				{
					mh2 = mh1;
					mh1 = clp.Item1;
				}
				else if (clp.Item1 > mh2)
				{
					mh2 = clp.Item1;
				}
			}
			return new Tuple<int, int>(Math.Max(mh1, mh2) + 1, Math.Max(childMax, mh1 + mh2 + 2));
		}
	}

	class Tree
	{
		public Tree(int node) : this(node, null) { }
		
		public Tree(int node, IEnumerable<Tree> children)
		{
			this.Node = node;
			this.Children = children;
		}

		public int Node { get; private set; }
		
		public IEnumerable<Tree> Children { get; private set; }
	}
}
